### ISSUE TITLE
Pure read workload latency problem

### ISSUE THREAD

**Issue by Xavier1994** [tikv/tikv#3911]
### Deploy
3 * TiKV(16C32G, 500G SSD), 1 * PD(8C16G, 200G SSD)

### Config
[readpool.storage]
high-concurrency = 32
low-concurrency = 32
max-tasks-per-worker-high = 10000
max-tasks-per-worker-low = 10000
max-tasks-per-worker-normal = 10000
normal-concurrency = 32

[readpool.coprocessor]
high-concurrency = 16
low-concurrency = 16
normal-concurrency = 16

### Problem
I have some problems about the read latency under pure read load, my dataset is 0.15 billion key value pairs and every pair's size is about 150 bytes, and every read request hits cache(cache hit is 100%). But when i add my read workload, the grpc latency became high. you can see, the local read request latency is almost the same,  but the grpc latency became higher when i add more read load 

![2018-12-10 11 46 31](https://user-images.githubusercontent.com/11145343/49744748-7623e700-fcd8-11e8-9f99-cd329695d9df.png)
![2018-12-10 11 47 08](https://user-images.githubusercontent.com/11145343/49744749-7623e700-fcd8-11e8-87ad-69a64e7cd122.png)
![2018-12-10 11 47 48](https://user-images.githubusercontent.com/11145343/49744750-76bc7d80-fcd8-11e8-834d-7d23a25a42d7.png)
![2018-12-10 11 49 23](https://user-images.githubusercontent.com/11145343/49744751-76bc7d80-fcd8-11e8-9cd0-9c0f3f55a502.png)

**Comment by overvenus** [comment]
Hi @Xavier1994  

> I have some problems about the read latency under pure read workload, my dataset is 0.15 billion key value pairs and every pair's size is about 150 bytes, and every read request hits cache(cache hit is 100%).

Could you paste kv rocksdb metrics?

> But when i add my read workload, the grpc latency became high. you can see, the local read request latency is almost the same, but the grpc latency became higher when i add more read load,

When did you increase read workload?

> I think it is unreasonable, So i want to know why and if i can fix this issue by change some configuration

A quick guess: requests may spend more time in channel(queueing, need further investigation).

**Comment by Xavier1994** [comment]
@overvenus 
the increase duration is short, so it is difficult to see from grafana, but you can see i decrease read workload at 23:13, the QPS became lower and the latency became lower. But local read request is always the same, and read related CPU and the network is not the bottleneck. 

![rocksdb](https://user-images.githubusercontent.com/11145343/49779293-3db9f280-fd44-11e8-8e2e-dc999da3db7a.jpg)

**Comment by Xavier1994** [comment]
Add storage metric panel

![2018-12-11 2 12 30](https://user-images.githubusercontent.com/11145343/49781738-da818d80-fd4e-11e8-8add-f4a3304d1d44.png)

**Comment by Xavier1994** [comment]
I add two more metrics

![2018-12-11 2 35 40](https://user-images.githubusercontent.com/11145343/49782692-313c9680-fd52-11e8-8c91-18567893aa2e.png)

Storage Raw Get Duration and Storage process read duration is the latency for variable _timer and _t_process

the related code is https://github.com/tikv/tikv/blob/d68649c760d136014fd39ee3ce5fcb7ec52df4f7/src/storage/mod.rs#L1010

**Comment by siddontang** [comment]
PTAL @overvenus

**Comment by Xavier1994** [comment]
I find the async_raw_get's latency = async_snapshot's latency + rocksdb get latency, and from the grafana panel, i can see async_snapshot's latency and rocksdb get latency is almost the same. but async raw get latency became higher as the read workload increases.  So i guess maybe it is because the read pool's thread waste time to do context switch, because read pool is cpu-intensive. so i bind read pool's thread to specific cpu core, and the result is as below

![2018-12-11 4 38 39](https://user-images.githubusercontent.com/11145343/49805281-b85c2f80-fd8f-11e8-980a-3e714811267c.png)

**Comment by siddontang** [comment]
> so i bind read pool's thread to specific cpu core

Very cool @Xavier1994 
How do you do this? 

Btw, I find that the QPS reduced at about 15:40 or later, what did you do at that time?

**Comment by breeswish** [comment]
Thanks for reporting. Are you using KV interface or Coprocessor interface? May you post the metrics of Tasks and Thread CPU?

**Comment by Xavier1994** [comment]
@siddontang I just use linux taskset to bind thread to specified cpu. And at 15:40, one of my benchmark program is ended. So the QPS reduced. 

@breeswish I use raw kv api(just use get and put operation),  thread cpu metrics haven posted above

**Comment by breeswish** [comment]
What do you mean by workload? Is it the concurrency of requests?

Also, please try adjust high/normal/low-concurrency to 3, which should help a lot reducing contention.

**Comment by Xavier1994** [comment]
Yes, workload means the concurrency of requests. I have three machines to run benchmark program, every benchmark program start 60 threads to send get requests to server. 

when I test, I have set high-concurrency and normal-concurrency to 1, coprocessor's high/normal/low was also set to 1, because my raw get request just use store-read-low threads. And i set low-concurrency to 8, grpc concurrency to 8. Then i get the data above.

**Comment by breeswish** [comment]
@Xavier1994 It looks to be normal and I do not have many ideas for you. I have not specifically checked about the latency increasing when there is more concurrency during my own benchmarks, but according to cpupool's implementation I think this can be possible. The best concurrency strategy we known so far is to set it equal to the real threads payload, hope latency can be more stable.
